File: Algorithms and Data Structures Enhancement/TaskService.cpp
#include "TaskService.h"
#include <iostream>
#include <algorithm>

// Helper method for comparing two strings to validate length and convert to lowercase
bool TaskService::equalsIgnoreCase(const std::string& a, const std::string& b) {
    // Check length of strings
    if (a.size() != b.size()) return false;
    // Increment through each character of string
    for (size_t i = 0; i < a.size(); ++i) {
        // Check if lowercase conversion has a mismatch
        if (std::tolower(static_cast<unsigned char>(a[i])) != std::tolower(static_cast<unsigned char>(b[i])))
            return false;
    }
    return true;
}

// Adds a new task if task list is empty
bool TaskService::addTask(std::shared_ptr<Task> task) {
    // Check for null task
    if (!task) return false;
    // If task list is empty, add task
    if (taskList.empty()) {
        taskList.push_back(task);
        return true;
    }
    // If task list is not empty but has unique task ID, add task to task list
    for (auto& c : taskList) {
        if (equalsIgnoreCase(task->getTaskId(), c->getTaskId())) {
            return false;
        }
    }
    taskList.push_back(task);
    return true;
}

// By utilizing task ID, return task object. If no task ID is found, return default value
std::shared_ptr<Task> TaskService::getTask(const std::string& taskId) const {
    for (auto& t : taskList) {
        if (t && t->getTaskId() == taskId) return t;
    }
    return nullptr;
}

// Delete task from task list by utilizing task ID
void TaskService::deleteTask(const std::string& taskId) {
    // Increment through each task in task list
    for (size_t i = 0; i < taskList.size(); ++i) {
        // If the task ID matches an existing task ID in the task list, delete the task
        if (taskList[i] && taskList[i]->getTaskId() == taskId) {
            taskList.erase(taskList.begin() + i);
            return;
        }
    }
    std::cout << "Task ID: " << taskId << " not found." << std::endl;
}

// Update task name utilizing task ID to find specific task
void TaskService::updateTaskName(const std::string& updatedString, const std::string& taskId) {
    // Increment through each task in task list
    for (size_t i = 0; i < taskList.size(); ++i) {
        // If the task ID matches an existing task ID in the task list, update task name
        if (taskList[i] && taskList[i]->getTaskId() == taskId) {
            taskList[i]->setTaskName(updatedString);
            return;
        }
    }
    std::cout << "Task ID: " << taskId << " not found." << std::endl;
}

// Update task description utilizing task ID to find specific task
void TaskService::updateTaskDescription(const std::string& updatedString, const std::string& taskId) {
    // Increment through each task in task list
    for (size_t i = 0; i < taskList.size(); ++i) {
        // If the task ID matches an existing task ID in the task list, update task description
        if (taskList[i] && taskList[i]->getTaskId() == taskId) {
            taskList[i]->setTaskDescription(updatedString);
            return;
        }
    }
    std::cout << "Task ID: " << taskId << " not found." << std::endl;
}

// Print the task ID, name, and description of each task in task list to console
void TaskService::displayTaskList() const {
    for (auto& t : taskList) {
        if (t) {
            std::cout << "Task ID: " << t->getTaskId() << "\n";
            std::cout << "Task Name: " << t->getTaskName() << "\n";
            std::cout << "Task Description: " << t->getTaskDescription() << "\n";
        }
    }
}

File: Algorithms and Data Structures Initial/src/Tests/TaskService.java
/*
 * Nicholas Feero
 * CS-320: Software Test & Automation
 * Module #4 Milestone | TaskService.java
 * Professor Toledo
 * February 2nd, 2025
*/

package Tests;

import java.util.ArrayList;

//creation of TaskService class
public class TaskService {
	//create an ArrayList to store tasks
	public ArrayList<Task> taskList = new ArrayList<Task>(0);
	//display full list of tasks to check for errors
	public void displayTaskList() {
		for (int i = 0; i < taskList.size(); i++) {
			System.out.println("Task ID: " + taskList.get(i).getTaskId());
			System.out.println("Task Name: " + taskList.get(i).getTaskName());
			System.out.println("Task Description: " + taskList.get(i).getTaskDescription());
		}
	}
	//adds a new task if task list is empty
	//checks task list if there's an identical task ID. If so, refuse addition of task
	//if task list isn't empty but has a unique task ID entry, add task
	public boolean addTask(Task task) {
		boolean didAdd = false;
		//list is empty
		if (taskList.size() == 0 ) {
			taskList.add(task);
			didAdd = true;
		}
		else {
			for(Task c : taskList) {
				if (task.getTaskId().equalsIgnoreCase(c.getTaskId())) {
					return didAdd;
				}
			} //end for loop
			taskList.add(task);
			didAdd = true;
		}
		return didAdd;
	}
	//by utilizing task ID, return task object. If no task ID is found, return default value
	public Task getTask(String taskId) {
		Task task = new Task(null, null, null);
		for (int i = 0; i < taskList.size(); i++) {
			if (taskList.get(i).getTaskId().contentEquals(taskId)) {
				task = taskList.get(i);
			}
		}
		return task;
	}
	//delete task method
	//utilize task ID to find correct task to delete
	//if list is traversed without a return of specified task ID, output no task ID found to console
	public void deleteTask(String taskId) {
		for (int i = 0; i < taskList.size(); i++) {
			if (taskList.get(i).getTaskId().equals(taskId)) {
				taskList.remove(i);
				break;
			}
			if (i == taskList.size() - 1) {
				System.out.println("Task ID: " + taskId + " not found.");
			}
		}
	}
	//update task name utilizing task ID to find specific task
	//if list is traversed without a return of specified task ID, output no task ID found to console
	public void updateTaskName(String updatedString, String taskId) {
		for (int i = 0; i < taskList.size(); i++) {
			if (taskList.get(i).getTaskId().equals(taskId)) {
				taskList.get(i).setTaskName(updatedString);
				break;
			}
			if (i == taskList.size() - 1) {
				System.out.println("Task ID: " + taskId + " not found.");
			}
		}
	}
	//update task description utilizing task ID to find specific task
	//if list is traversed without a return of specified task ID, output no task ID found to console
	public void updateTaskDescription(String updatedString, String taskId) {
		for (int i = 0; i < taskList.size(); i++) {
			if (taskList.get(i).getTaskId().equals(taskId)) {
				taskList.get(i).setTaskDescription(updatedString);
				break;
			}
			if (i == taskList.size() - 1) {
				System.out.println("Task ID: " + taskId + " not found.");
			}
		}
	}	
}
